purchases/src/main/kotlin/com/revenuecat/purchases/google/ProductValidation.kt

```diff
@@ -0,0 +1,77 @@
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.PurchasesErrorCode
+import com.revenuecat.purchases.PurchasesException
+import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.models.GooglePurchasingData
+import com.revenuecat.purchases.models.Period
+import com.revenuecat.purchases.models.PurchasingData
+import com.revenuecat.purchases.models.StoreProduct
+import kotlin.jvm.Throws
+
+/**
+ * Maximum number of add-on products allowed in a multi-line purchase.
+ */
+internal const val MAX_NUMBER_OF_ADD_ON_PRODUCTS = 49
+
+/**
+ * Validates and filters add-on products for multi-line purchases.
+ *
+ * Performs compatibility checks and filters add-ons to ensure they are compatible with the base product.
+ *
+ * @param baseProductPurchasingData The purchase data for primary product that add-ons will be bundled with
+ * @param addOnProducts The list of potential add-on products to validate and filter
+ * @return A filtered list containing only add-ons that match the base product's purchasing data type
+ * @throws PurchasesException If the add-ons failed validation for some reason.
+ */
+@ExperimentalPreviewRevenueCatPurchasesAPI
+@Throws(PurchasesException::class)
+@Suppress("ThrowsCount")
+internal fun validateAndFilterCompatibleAddOnProducts(
+    baseProductPurchasingData: PurchasingData,
+    addOnProducts: List<StoreProduct>,
+): List<GooglePurchasingData> {
+    if (baseProductPurchasingData !is GooglePurchasingData.Subscription) {
+        val error = PurchasesError(
+            PurchasesErrorCode.PurchaseInvalidError,
+            "Add-ons are currently only supported for Google subscriptions.",
+        ).also { errorLog(it) }
+
+        throw PurchasesException(error)
+    }
+    val addOnProductsWithSameProductType: MutableList<GooglePurchasingData> = ArrayList()
+    val billingPeriods = mutableSetOf<Period>()
+
+    for (addOnProduct in addOnProducts) {
+        val addOnPurchasingData = addOnProduct.purchasingData
+        if (addOnPurchasingData is GooglePurchasingData &&
+            addOnPurchasingData::class == baseProductPurchasingData::class
+        ) {
+            addOnProductsWithSameProductType.add(addOnPurchasingData)
+            addOnProduct.period?.let { billingPeriods.add(it) }
+        }
+    }
+
+    if (addOnProductsWithSameProductType.count() > MAX_NUMBER_OF_ADD_ON_PRODUCTS) {
+        val error = PurchasesError(
+            PurchasesErrorCode.PurchaseInvalidError,
+            "Multi-line purchases cannot contain more than ${MAX_NUMBER_OF_ADD_ON_PRODUCTS + 1} products " +
+                "(1 base + $MAX_NUMBER_OF_ADD_ON_PRODUCTS add-ons).",
+        ).also { errorLog(it) }
+
+        throw PurchasesException(error)
+    }
+
+    if (billingPeriods.size > 1) {
+        val error = PurchasesError(
+            PurchasesErrorCode.PurchaseInvalidError,
+            "All items in a multi-line purchase must have the same billing period.",
```

Comment:
Just curious, what happens if one of the add-on products has a free trial/intro price with a different period? Does that get ignored?

Reply:
That is a _great_ question. I'll test this out today and see! If it causes an issue, I'll add a check here for it.